File: kws_de/stream.py
from collections import deque

import numpy as np


class KeywordStream:
    """Smooths per-step posteriors and decodes them into discrete keyword
    events using edge-triggered, run-based decoding.

    Each step's *candidate* is the top-1 label of the smoothed posterior
    (trailing mean over `smooth_win` raw pushes) if its probability is
    >= `threshold` and it isn't `_silence_`, else `None`. Consecutive steps
    with the same candidate form a *run*; a run fires its label exactly
    once, as soon as it reaches `min_consecutive` steps (never again while
    that same candidate persists — no global cooldown, so a different
    label's run can fire immediately after its own run qualifies). A run
    of the SAME label that most recently fired may fire again only after
    at least `gap_steps` consecutive non-matching steps have elapsed since
    its previous run ended.
    """

    def __init__(
        self, predict_fn, labels, smooth_win=3, threshold=0.5, min_consecutive=2, gap_steps=2
    ):
        self.predict_fn = predict_fn
        self.labels = list(labels)
        self.smooth_win = smooth_win
        self.threshold = threshold
        self.min_consecutive = min_consecutive
        self.gap_steps = gap_steps
        self.reset()

    def reset(self):
        self._hist = deque(maxlen=self.smooth_win)
        self._run_label = None
        self._run_len = 0
        self._run_fired = False
        self._last_fired_label = None
        self._gap_since_last_fired = 0

    def push(self, posterior) -> list:
        self._hist.append(np.asarray(posterior, dtype=np.float64))
        smoothed = np.mean(self._hist, axis=0)
        idx = int(np.argmax(smoothed))
        label = self.labels[idx]
        candidate = label if (smoothed[idx] >= self.threshold and label != "_silence_") else None

        if candidate == self._run_label:
            self._run_len += 1
        else:
            self._run_label = candidate
            self._run_len = 1
            self._run_fired = False

        events = []
        if candidate is not None and self._run_len >= self.min_consecutive and not self._run_fired:
            gap_ok = (
                candidate != self._last_fired_label or self._gap_since_last_fired >= self.gap_steps
            )
            if gap_ok:
                events.append(candidate)
                self._run_fired = True
                self._last_fired_label = candidate
                self._gap_since_last_fired = 0

        if self._last_fired_label is not None and candidate != self._last_fired_label:
            self._gap_since_last_fired += 1

        return events

File: firmware/main/stream.c
#include "stream.h"
#include <string.h>

void stream_reset(stream_t *s)
{
    memset(s, 0, sizeof *s);
    s->run_label = -1;
    s->last_fired_label = -1;
}

int stream_push(stream_t *s, const float posterior[KWS_NUM_LABELS])
{
    memcpy(s->hist[s->hist_pos], posterior, sizeof(float) * KWS_NUM_LABELS);
    s->hist_pos = (s->hist_pos + 1) % KWS_SMOOTH_WIN;
    if (s->hist_len < KWS_SMOOTH_WIN) s->hist_len++;

    float best = -1.f; int idx = 0;
    for (int k = 0; k < KWS_NUM_LABELS; k++) {
        float m = 0.f;
        for (int h = 0; h < s->hist_len; h++) m += s->hist[h][k];
        m /= s->hist_len;
        if (m > best) { best = m; idx = k; }
    }
    int candidate = (best >= KWS_THRESHOLD && idx != KWS_SILENCE_INDEX) ? idx : -1;

    if (candidate == s->run_label) s->run_len++;
    else { s->run_label = candidate; s->run_len = 1; s->run_fired = 0; }

    int fired = -1;
    if (candidate >= 0 && s->run_len >= KWS_MIN_CONSECUTIVE && !s->run_fired) {
        int gap_ok = candidate != s->last_fired_label || s->gap_since_last_fired >= KWS_GAP_STEPS;
        if (gap_ok) {
            fired = candidate;
            s->run_fired = 1;
            s->last_fired_label = candidate;
            s->gap_since_last_fired = 0;
        }
    }
    if (s->last_fired_label >= 0 && candidate != s->last_fired_label) s->gap_since_last_fired++;
    return fired;
}
